File: backend/src/main/java/com/comfortableledger/ledger/util/SecretValues.java
package com.comfortableledger.ledger.util;

public final class SecretValues {
    private static final String MASK = "********";

    private SecretValues() {
    }

    public static String mask(String value) {
        String normalized = StringValues.trimToEmpty(value);
        if (normalized.isBlank()) {
            return "";
        }
        if (normalized.length() <= 4) {
            return MASK;
        }
        return normalized.substring(0, 2) + MASK + normalized.substring(normalized.length() - 2);
    }

    public static String maskAll(String value) {
        return StringValues.trimToEmpty(value).isBlank() ? "" : MASK;
    }
}

File: frontend/src/utils/secretValues.js
const MASK = '********';

export function maskSecret(value) {
  const normalized = String(value ?? '').trim();
  if (!normalized) return '';
  if (normalized.length <= 4) return MASK;
  return `${normalized.slice(0, 2)}${MASK}${normalized.slice(-2)}`;
}

export function maskAll(value) {
  return String(value ?? '').trim() ? MASK : '';
}
